src/v/kafka/server/usage_aggregator.cc

```diff
@@ -117,6 +117,31 @@ auto epoch_time_secs(time_point now) {
       .count();
 }
 
+static ss::lowres_system_clock::time_point round_to_interval(
+  std::chrono::seconds usage_window_width_interval,
+  ss::lowres_system_clock::time_point t) {
+    /// Downstream systems are particularly sensitive to minor issues with
+    /// timestamps not triggering on the configured interval (hours, minutes,
+    /// seconds, etc), this method rounds t to the nearest interval and logs an
+    /// error if this cannot be done within some threshold.
+    static const auto err_threshold = 5s;
+    const auto diff_secs = std::chrono::seconds(
+      epoch_time_secs(t) % usage_window_width_interval.count());
+    const auto ts = t - diff_secs;
+    const auto min_ts = t - err_threshold;
+    const auto max_ts = t + err_threshold;
+    if (min_ts > ts || ts > max_ts) {
```

Comment:
I'm not sure what the expected behaviour is here if we have backwards clock drift (i.e. `t < usage_window_width_interval`). The comment above states that `t` should be rounded to the nearest interval, but the code actually rounds to the smallest interval boundary smaller than `t`. For example, if `t=95` and `usage_window_width_interval=100`, then `ts=0` and the function still returns `95`.

Reply:
Since `t` is a system clock time point, the above could actually happen.

Reply:
Yeah good point that wasn't my intention +1

Reply:
fixed